peppy/project.py

```diff
@@ -473,23 +481,63 @@ def _auto_merge_duplicated_names(self):
                 f"you may use either auto-merging or subsample_table-based merging. "
                 f"Duplicates: {dups_set}"
             )
-        for dup in dups_set:
-            dup_samples = [s for s in self.samples if getattr(s, self.st_index) == dup]
+        for duplication in dups_set:
+            (
+                duplicated_samples,
+                non_duplicated_samples,
+            ) = self._get_duplicated_and_not_duplicated_samples(
+                duplication, self.st_index, self.samples
+            )
+            self._samples = non_duplicated_samples
+
             sample_attrs = [
-                attr for attr in dup_samples[0].keys() if not attr.startswith("_")
+                attr
+                for attr in duplicated_samples[0].keys()
+                if not attr.startswith("_")
             ]
+
             merged_attrs = {}
             for attr in sample_attrs:
                 merged_attrs[attr] = list(
-                    flatten([getattr(s, attr) for s in dup_samples])
+                    flatten([getattr(s, attr) for s in duplicated_samples])
                 )
+
             # make single element lists scalars
-            for k, v in merged_attrs.items():
-                if isinstance(v, list) and len(list(set(v))) == 1:
-                    merged_attrs[k] = v[0]
-            self._samples = [s for s in self._samples if s[self.st_index] != dup]
+            for attribute, values in merged_attrs.items():
+                if isinstance(
+                    values, list
+                ) and self._all_values_in_the_list_are_the_same(values):
+                    merged_attrs[attribute] = values[0]
+
             self.add_samples(Sample(series=merged_attrs))
 
+    @staticmethod
+    def _get_duplicated_and_not_duplicated_samples(
+        duplication: str, st_index: str, samples: List[Sample]
+    ) -> Tuple[List, List]:
+        """
+        Iterates over list of samples and splits them into list of duplicated samples and list of not duplicated.
+
+        Args:
+            duplication: Sample ID.
+            st_index: Sample table index.
+            samples: List of Sample instances.
+
+        Returns:
+            List of Sample objects of duplicated samples and not duplicated samples.
+        """
+        duplicated_samples, not_duplicated_samples = [], []
+        for sample in samples:
+            if sample[st_index] == duplication:
+                duplicated_samples.append(sample)
+            else:
+                not_duplicated_samples.append(sample)
+        return duplicated_samples, not_duplicated_samples
+
+    @staticmethod
+    def _all_values_in_the_list_are_the_same(list_of_values: List) -> bool:
+        return all(value == list_of_values[0] for value in list_of_values) == 1
```

Comment:
Does `all()` return a `bool`? Do we need the ` == 1`?. Also, not sure it matters, but reading [here](https://stackoverflow.com/questions/3844801/check-if-all-elements-in-a-list-are-identical), using `itertools.groupby` might be faster? I'm sure you've already read this thread, however.

Reply:
You are right, we probably don't need `==1`. As to the method used it looks like both of them present similar performance 
![image](https://user-images.githubusercontent.com/43926522/179570858-40f73e2b-cf04-4e13-bbb3-a8be0aa25259.png)